MMSS: Connection Logic Changes (#11519)
nrfconnect/sdk-nrf: samples/nrf9160/nrf_cloud_mqtt_multi_service/README.rst

The top level functional unit and entry point for the sample is the :file:`src/main.c` file.
This file starts three primary threads, each with a distinct function:

* The cloud connection thread (``con_thread``, :file:`src/cloud_connection.c`) runs the :ref:`nrf_cloud_mqtt_multi_service_cloud_connection_loop`, which maintains a connection to `nRF Cloud`_.

> * The application thread (``app_thread``, :file:`src/application.c`) runs the :ref:`nrf_cloud_mqtt_multi_service_application_thread_and_main_application_loop`, which controls demo features, and submits `device messages <nRF Cloud Device Messages_>`_ to the :ref:`nrf_cloud_mqtt_multi_service_device_message_queue`.

Reviewer: ```suggestion
* The application thread (``app_thread``, :file:`src/application.c`) runs the :ref:`nrf_cloud_mqtt_multi_service_application_thread_and_main_application_loop`, which controls demo features and submits `device messages <nRF Cloud Device Messages_>`_ to the :ref:`nrf_cloud_mqtt_multi_service_device_message_queue`.
```

> * The message queue thread (``msg_thread``, :file:`src/message_queue.c`) then transmits these messages whenever there is an active connection. (See :ref:`nrf_cloud_mqtt_multi_service_device_message_queue`).

Reviewer: ```suggestion
* The message queue thread (``msg_thread``, :file:`src/message_queue.c`) then transmits these messages whenever there is an active connection.
  See :ref:`nrf_cloud_mqtt_multi_service_device_message_queue`.
```

:file:`src/main.c` also optionally starts a fourth thread, the ``led_thread``, which animates any onboard LEDs if :ref:`nrf_cloud_mqtt_multi_service_led_status_indication` is enabled.

.. _nrf_cloud_mqtt_multi_service_cloud_connection_loop:

Cloud connection loop
=====================

> The cloud connection loop (implemented in :file:`src/cloud_connection.c`) monitors network availability, and starting a connection with `nRF Cloud`_ whenever the internet becomes reachable, and closing that connection whenever internet access is lost.

Reviewer: ```suggestion
The cloud connection loop (implemented in :file:`src/cloud_connection.c`) monitors network availability.
It starts a connection with `nRF Cloud`_ whenever the Internet becomes reachable, and closes that connection whenever Internet access is lost.
```

> It has error handling and timeout features to ensure failed or lost connections are re-established after a waiting period (:kconfig:option:`CONFIG_CLOUD_CONNECTION_RETRY_TIMEOUT_SECONDS`).

Reviewer: ```suggestion
It has error handling and timeout features to ensure that failed or lost connections are re-established after a waiting period (:kconfig:option:`CONFIG_CLOUD_CONNECTION_RETRY_TIMEOUT_SECONDS`).
```

LTE is enabled and connected to automatically by Zephyr's ``conn_mgr`` and :kconfig:option:`CONFIG_LTE_CONNECTIVITY`.

Whenever a connection to nRF Cloud is started, the cloud connection loop follows the :ref:`nRF Cloud connection process <lib_nrf_cloud_connect>`.

> The :ref:`lib_nrf_cloud` library handles most of the connection process, with exception to the following behavior:

Reviewer: ```suggestion
The :ref:`lib_nrf_cloud` library handles most of the connection process, with exception to the following behavior:

```

When a device is first being associated with an nRF Cloud account, the `nRF Cloud MQTT API`_ will send a user association request notification to the device.
Upon receiving this notification, the device must wait for a user association success notification, and then manually disconnect from and reconnect to nRF Cloud.
Notifications of user association success are sent for every subsequent connection after this as well, so the device must only disconnect and reconnect if it previously received a user association request notification.

> This behavior is handled by the ``NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST`` and ``NRF_CLOUD_EVT_USER_ASSOCIATED`` cases inside the :c:func:`cloud_event_handler` function.

Reviewer: Do you really call them "cases". Are they not rather events or something like that?

Author: Yep -- The events are handled by the corresponding cases in a switch statement, so when referring to the event handler, it is better to say case

Upon startup, the cloud connection loop also updates the `Device Shadow <nRF Cloud Device Shadows_>`_.
This is performed in the :c:func:`update_shadow` function.

.. _nrf_cloud_mqtt_multi_service_device_message_queue:

Device message queue
====================

Any thread may submit `device messages <nRF Cloud Device Messages_>`_ to the device message queue (implemented in :file:`src/message_queue.c`), where they are stored until a working connection to `nRF Cloud`_ is established.
Once this happens, the message thread transmits all enqueued device messages, one at a time and in fast succession, to nRF Cloud.
If an enqueued message fails to send, it will be sent back to the queue and tried again later.
Transmission is paused whenever connection to nRF Cloud is lost.
If more than :ref:`CONFIG_MAX_CONSECUTIVE_SEND_FAILURES <CONFIG_MAX_CONSECUTIVE_SEND_FAILURES>` messages fail to send in a row, the connection to nRF Cloud is reset, and then reconnected after a delay (:kconfig:option:`CONFIG_CLOUD_CONNECTION_RETRY_TIMEOUT_SECONDS`).

> Most messages sent by this sample to nRF Cloud are sent using the message queue, but the following are instead sent directly:

Reviewer: ```suggestion
Most messages sent by this sample to nRF Cloud are sent using the message queue, but the following are sent directly:
```

This file starts three primary threads, each with a distinct function:

* The cloud connection thread (``con_thread``, :file:`src/cloud_connection.c`) runs the :ref:`nrf_cloud_mqtt_multi_service_cloud_connection_loop`, which maintains a connection to `nRF Cloud`_.
* The application thread (``app_thread``, :file:`src/application.c`) runs the :ref:`nrf_cloud_mqtt_multi_service_application_thread_and_main_application_loop`, which controls demo features and submits `device messages <nRF Cloud Device Messages_>`_ to the :ref:`nrf_cloud_mqtt_multi_service_device_message_queue`.
* The message queue thread (``msg_thread``, :file:`src/message_queue.c`) then transmits these messages whenever there is an active connection.
  See :ref:`nrf_cloud_mqtt_multi_service_device_message_queue`.

:file:`src/main.c` also optionally starts a fourth thread, the ``led_thread``, which animates any onboard LEDs if :ref:`nrf_cloud_mqtt_multi_service_led_status_indication` is enabled.

.. _nrf_cloud_mqtt_multi_service_cloud_connection_loop:

Cloud connection loop
=====================

The cloud connection loop (implemented in :file:`src/cloud_connection.c`) monitors network availability.

> It starts a connection with `nRF Cloud`_ whenever the internet becomes reachable, and closes that connection whenever Internet access is lost.

Reviewer: ```suggestion
It starts a connection with `nRF Cloud`_ whenever the Internet becomes reachable, and closes that connection whenever Internet access is lost.
```

Since the :kconfig:option:`CONFIG_LTE_CONNECTIVITY` Kcofig option is enabled, Zephyr's ``conn_mgr`` automatically enables and connects to LTE.

Whenever a connection to nRF Cloud is started, the cloud connection loop follows the :ref:`nRF Cloud connection process <lib_nrf_cloud_connect>`.
The :ref:`lib_nrf_cloud` library handles most of the connection process, with exception to the following behavior:

When a device is first being associated with an nRF Cloud account, the `nRF Cloud MQTT API`_ will send a user association request notification to the device.
Upon receiving this notification, the device must wait for a user association success notification, and then manually disconnect from and reconnect to nRF Cloud.
Notifications of user association success are sent for every subsequent connection after this as well, so the device must only disconnect and reconnect if it previously received a user association request notification.
This behavior is handled by the ``NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST`` and ``NRF_CLOUD_EVT_USER_ASSOCIATED`` cases inside the :c:func:`cloud_event_handler` function.

> Upon startup, the cloud connection loop also updates the `Device Shadow <nRF Cloud Device Shadows_>`_.

Reviewer: ```suggestion
Upon startup, the cloud connection loop also updates the `device shadow <nRF Cloud Device Shadows_>`_.
```

This is performed in the :c:func:`update_shadow` function.

.. _nrf_cloud_mqtt_multi_service_device_message_queue:

Device message queue
====================

Any thread may submit `device messages <nRF Cloud Device Messages_>`_ to the device message queue (implemented in :file:`src/message_queue.c`), where they are stored until a working connection to `nRF Cloud`_ is established.
Once this happens, the message thread transmits all enqueued device messages, one at a time and in fast succession, to nRF Cloud.
If an enqueued message fails to send, it will be sent back to the queue and tried again later.
Transmission is paused whenever connection to nRF Cloud is lost.

> If more than :ref:`CONFIG_MAX_CONSECUTIVE_SEND_FAILURES <CONFIG_MAX_CONSECUTIVE_SEND_FAILURES>` messages fail to send in a row, the connection to nRF Cloud is reset, and then reconnected after a delay (:kconfig:option:`CONFIG_CLOUD_CONNECTION_RETRY_TIMEOUT_SECONDS`).

Reviewer: ```suggestion
If more than :ref:`CONFIG_MAX_CONSECUTIVE_SEND_FAILURES <CONFIG_MAX_CONSECUTIVE_SEND_FAILURES>` messages in a row fail to send, the connection to nRF Cloud is reset, and then reconnected after a delay (:kconfig:option:`CONFIG_CLOUD_CONNECTION_RETRY_TIMEOUT_SECONDS`).
```

> Most messages sent by this sample to nRF Cloud are sent using the message queue, but the following are sent directly:

Reviewer: ```suggestion
Most messages sent to nRF Cloud by this sample are sent using the message queue, but the following are sent directly:
```

* Alerts using the :ref:`lib_nrf_cloud_alert` library.
* Logs using the :ref:`lib_nrf_cloud_log` library.

> * `Device Shadow <nRF Cloud Device Shadows_>`_ updates.

Reviewer: ```suggestion
* `Device shadow <nRF Cloud Device Shadows_>`_ updates.
```

* Constructs timestamped sensor sample and location `device messages <nRF Cloud Device Messages_>`_ using `cJSON`_.
* Sends sensor sample and location device messages to the :ref:`nrf_cloud_mqtt_multi_service_device_message_queue`.
* Checks for and executes :ref:`remote modem AT command requests <nrf_cloud_mqtt_multi_service_remote_at>`.

> * Sends temperature and sample start alerts using the :ref:`lib_nrf_cloud_alert` library.

Reviewer: ```suggestion
* Sends the temperature and sample start alerts using the :ref:`lib_nrf_cloud_alert` library.
```
Not sure I understand this correctly. Does the thread send "sample start" alerts? Or does the sample start generating alerts?

Author: It sends "temperature alerts" and  "sample start alert(s)". The phrasing is admittedly hard to follow, and whether "sample start alerts" should be plural is especially unclear to me since exactly one is sent, but it is sent every time the sample boots.

Perhaps:

`Sends temperature alerts and a sample startup alerts using the...` would be clearest. I've switched to that, let me know if it is acceptable
